pkg/controller/certificaterequests/sync.go

```diff
@@ -0,0 +1,239 @@
+	if err != nil {
+		return err
+	}
+
+	el := validation.ValidateCertificateRequest(crCopy)
+	if len(el) > 0 {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, "BadConfig", "Resource validation failed: %v", el.ToAggregate())
+		return nil
+	}
+
+	// step zero: check if the referenced issuer exists and is ready
+	issuerObj, err := c.helper.GetGenericIssuer(crCopy.Spec.IssuerRef, crCopy.Namespace)
+	if k8sErrors.IsNotFound(err) {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerNotFound, err.Error())
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	dbg.Info("Fetched issuer resource referenced by certificate request", "issuer_name", crCopy.Spec.IssuerRef.Name)
+
+	issuerReady := apiutil.IssuerHasCondition(issuerObj, v1alpha1.IssuerCondition{
+		Type:   v1alpha1.IssuerConditionReady,
+		Status: v1alpha1.ConditionTrue,
+	})
+	if !issuerReady {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerNotReady, "Issuer %s not ready", issuerObj.GetObjectMeta().Name)
+		return nil
+	}
+
+	i, err := c.issuerFactory.IssuerFor(issuerObj)
+	if err != nil {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerInit, "Internal error initialising issuer: %v", err)
+		return nil
+	}
+
+	el = validation.ValidateCertificateRequestForIssuer(crCopy, issuerObj)
+	if len(el) > 0 {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, "BadConfig", "Resource validation failed: %v", el.ToAggregate())
+		return nil
+	}
+
+	if len(cr.Status.Certificate) == 0 {
+		dbg.Info("Invoking sign function as existing certificate does not exist")
+		return c.sign(ctx, i, crCopy, csr)
+	}
+
+	cert, err := pki.DecodeX509CertificateBytes(cr.Status.Certificate)
+	if err != nil {
+		return err
+	}
+
+	dbg.Info("Update certificate request status if required")
+	c.setCertificateRequestStatus(crCopy, csr, cert)
+
+	// TODO: Metrics??
+
+	if cert == nil {
+		dbg.Info("Invoking sign function as existing certificate does not exist")
+		return c.sign(ctx, i, crCopy, csr)
+	}
+
+	dbg.Info("Certificate does not need updating.")
+
+	return nil
+}
+
+// return an error on failure. If retrieval is succesful, the certificate data
+// will be stored in the certificate request status
+func (c *controller) sign(ctx context.Context, issuer issuer.Interface,
+	cr *v1alpha1.CertificateRequest, csr *x509.CertificateRequest) error {
+	log := logf.FromContext(ctx)
+
+	resp, err := issuer.Sign(ctx, cr)
+	if err != nil {
+		log.Error(err, "error issuing certificate request")
+		return err
+	}
+
+	// if the issuer has not returned any data, exit early
+	if resp == nil {
+		c.setCertificateRequestStatus(cr, nil, nil)
+		return nil
+	}
+
+	if len(resp.Certificate) > 0 {
+		cr.Status.Certificate = resp.Certificate
+		cr.Status.CA = resp.CA
+
+		c.recorder.Event(cr, corev1.EventTypeNormal, successCertificateIssued, "Certificate issued successfully")
+	}
+
+	cert, err := pki.DecodeX509CertificateBytes(resp.Certificate)
+	if err != nil {
+		return err
```

Comment:
I think we need to be more considered about returning an error here again. I think it may make sense for us to more universally (across all controllers) have some form of error handling function in the base controller, that can detect different types of error and conditionally retry the item if it makes sense to.